runtime/major_gc.c

```diff
@@ -562,10 +581,164 @@ static void mark_ephe_aux (struct mark_stack *stk, intnat *work,
   }
 }
 
+
+#define Pb_size (1 << 8)
+#define Pb_min 64
+#define Pb_mask (Pb_size - 1)
+
+Caml_inline void prefetch_block(value v)
+{
+  /* Prefetch a block so that scanning it later avoids cache misses.
+     We will access at least the header, but we don't yet know how
+     many of the fields we will access - the block might be already
+     marked, not scannable, or very short. The compromise here is to
+     prefetch the header and the first few fields.
+
+     We issue two prefetches, with the second being a few words ahead
+     of the first. Most of the time, these will land in the same
+     cacheline, be coalesced by hardware, and so not cost any more
+     than a single prefetch. Two memory operations are issued only
+     when the two prefetches land in different cachelines.
+
+     In the case where the block is not already in cache, and yet is
+     already marked, not markable, or extremely short, then we waste
+     somewhere between 1/8-1/2 of a prefetch operation (in expectation,
+     depending on alignment, word size, and cache line size), which is
+     cheap enough to make this worthwhile. */
+  caml_prefetch(Hp_val(v));
+  caml_prefetch(&Field(v, 3));
+}
+
+Caml_inline uintnat rotate1(uintnat x)
+{
+  return (x << ((sizeof x)*8 - 1)) | (x >> 1);
+}
+
+Caml_noinline static intnat do_some_marking
+#ifndef CAML_INSTR
+  (intnat work)
+#else
+  (intnat work, int* slice_fields, int* slice_pointers)
+#endif
+{
+  uintnat pb_enqueued = 0, pb_dequeued = 0;
+  int darkened_anything = 0;
+  value pb[Pb_size];
+  /* These global values are cached in locals,
+     so that they can be stored in registers */
+  struct mark_stack stk = *Caml_state->mark_stack;
+  uintnat young_start = (uintnat)Val_hp(Caml_state->young_start);
+  uintnat half_young_len =
+    ((uintnat)Caml_state->young_end - (uintnat)Caml_state->young_start) >> 1;
+#define Is_block_and_not_young(v) \
+  (((intnat)rotate1((uintnat)v - young_start)) > (intnat)half_young_len)
+#ifdef NO_NAKED_POINTERS
+  #define Is_major_block(v) Is_block_and_not_young(v)
```

Comment:
`Is_major_block` is essentially the same as `Is_block_and_not_young && Is_in_heap_or_young` (no conditional definition).

Reply:
I think I prefer the conditional definition, because it really is checking a different condition in no-naked-pointers mode. (as you've observed, it examines static data only in no-naked-pointers mode)

Reply:
To be clear, `Is_in_heap_or_young` is really a no-op in no-naked-pointers mode (but maybe it is poorly named). The conditional definition is already inside `Is_in_heap_or_young`.